marconi/src/Marconi/Index/MintBurn.hs

```diff
@@ -0,0 +1,310 @@
+
+    toRows :: TxMintEvent -> [[SQL.SQLData]]
+    toRows e = do
+      (txId, txMintAssets) <- NE.toList $ txMintEventTxAssets e
+      mintAsset <- NE.toList txMintAssets
+      pure
+        [ SQL.toField $ txMintEventSlotNo e
+        , SQL.toField $ txMintEventBlockHeaderHash e
+        , SQL.toField txId
+        , SQL.toField $ mintAssetPolicyId mintAsset
+        , SQL.toField $ mintAssetAssetName mintAsset
+        , SQL.toField $ mintAssetQuantity mintAsset
+        , SQL.toField $ mintAssetRedeemerIdx mintAsset
+        , SQL.toField $ mintAssetRedeemerData mintAsset
+        ]
+
+type Row = (C.SlotNo, C.Hash C.BlockHeader, C.TxId, C.PolicyId, C.AssetName, C.Quantity, Word64, C.ScriptData)
+
+-- | Input rows must be sorted by C.SlotNo.
+fromRows :: [Row] -> [TxMintEvent]
+fromRows rows =  do
+  rs@(r :| _) <- NE.groupBy ((==) `on` slotNo) rows -- group by SlotNo
+  pure $ TxMintEvent (slotNo r) (hash r) $ do
+    rs' <- NE.groupBy1 ((==) `on` txId) rs -- group by TxId
+    pure (txId r, rowToMintAsset <$> rs')
+  where
+    slotNo = view L._1 :: Row -> C.SlotNo
+    hash = view L._2 :: Row -> C.Hash C.BlockHeader
+    txId = view L._3 :: Row -> C.TxId
+    rowToMintAsset :: Row -> MintAsset
+    rowToMintAsset row = MintAsset (row^.L._4) (row^.L._5) (row^.L._6) (row^.L._7) (row^.L._8)
+
+sqliteSelectByTxIdPolicyId :: SQL.Connection -> (SQL.Query, [NamedParam]) -> C.TxId -> C.PolicyId -> IO [TxMintEvent]
+sqliteSelectByTxIdPolicyId sqlCon (conditions, params) txId policyId =
+  fromRows <$> SQL.queryNamed sqlCon query ([":txId" := txId, ":policyId" := policyId] <> params)
+  where query =
+          " SELECT slotNo, blockHeaderHash, txId, policyId                    \
+          \      , assetName, quantity, redeemerIx, redeemerData              \
+          \   FROM minting_policy_events                                      \
+          \  WHERE txId = :txId AND policyId = :policyId " <> conditions <> " \
+          \  ORDER BY slotNo, txId                                            "
+
+sqliteSelectAll :: SQL.Connection -> (SQL.Query, [NamedParam]) -> IO [TxMintEvent]
+sqliteSelectAll sqlCon (conditions, params) = fromRows <$> SQL.queryNamed sqlCon query params
+  where
+    whereClause = if conditions == "" then "" else " WHERE " <> conditions <> " "
+    query =
+      "   SELECT slotNo, blockHeaderHash, txId, policyId       \
+      \        , assetName, quantity, redeemerIx, redeemerData \
+      \     FROM minting_policy_events                         \
+      \     " <> whereClause <> "                              \
+      \ ORDER BY slotNo, txId                                  "
+
+intervalToWhereClause :: RI.QueryInterval C.ChainPoint -> (SQL.Query, [NamedParam])
+intervalToWhereClause qi = case qi of
+  RI.QEverything -> ("", [])
+  RI.QInterval from to
+    | Just from' <- slotMaybe from, Just to' <- slotMaybe to -> ("slotNo BETWEEN :fromSlot AND :toSlot" , [":fromSlot" := from', ":toSlot" := to'])
```

Comment:
This is not how intervals actually work. The condition for the persistent layer should be that the slot < end (where end is the top of the interval).

Reference: https://github.com/input-output-hk/plutus-apps/blob/d9eebba0e5b7f3f0d955e9755c7a031eb4de1303/rewindable-index/src/RewindableIndex/Storable.hs#L318-L329

Reply:
Yup, the comment I read several times without added clarity. So I want all events from the very beginning up to the latest point between start and end?